sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -290,6 +290,21 @@ object OptimizeIn extends Rule[LogicalPlan] {
  * 4. Removes `Not` operator.
  */
 object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
+  // Given argument x, return true if expression Not(x) can be simplified
+  // E.g. when x == Not(y), Not(x) == Not(Not(y)) == y
+  // For the case of x = EqualTo(a, b), recursively check each child expression
+  // Extra nullable check is required for EqualNullSafe because
+  // Not(EqualNullSafe(x, null)) is different from EqualNullSafe(x, Not(null))
+  private def canSimplifyNot(x: Expression): Boolean = x match {
+    case Literal(_, BooleanType) | Literal(_, NullType) => true
+    case _: Not | _: IsNull | _: IsNotNull | _: And | _: Or => true
+    case _: GreaterThan | _: GreaterThanOrEqual | _: LessThan | _: LessThanOrEqual => true
+    case EqualTo(a, b) if canSimplifyNot(a) || canSimplifyNot(b) => true
```

Comment:
I'm not sure we should recursive into `EqualTo`. For example:
`Not(EqualTo(a, EqualTo(b, EqualTo(c, d))))`, if `d` is optimizable, we will turn it into `EqualTo(Not(a), EqualTo(Not(b), EqualTo(Not(c), optimized_d)))`. We need to evaluate `Not` more times which is a perf degradation.

Reply:
Actually the code does not work in the way you described.
It works like this.
`Not(EqualTo(a, EqualTo(b, EqualTo(c, d))))` where `d` is optimizable, first `canSimplifyNot(a)` or `canSimplifyNot(EqualTo(b, EqualTo(c, d)))` are called depending on the case matches here
https://github.com/apache/spark/pull/33930/files/f5d3984c04029bdb7f4f00ef4de25a9e9fbc75cb#diff-d43484d56a4d9991066b5c00d12ec2465c75131e055fc02ee7fb6dfd45b5006fR472-R473
If `a` happens to be optimizable, `canSimplifyNot(a)` returns `true` and it propagates `Not` to `a`.
Otherwise, `canSimplifyNot(EqualTo(b, EqualTo(c, d)))` will call `canSimplifyNot(b)` and `canSimplifyNot(EqualTo(c, d))`, then that will call `canSimplifyNot(c)` and `canSimplifyNot(d)`.
Since `d` is optimizable in this example, `canSimplifyNot(d)` returns `true` that makes `canSimplifyNot(EqualTo(c, d))` return `true`. Eventually `canSimplifyNot(EqualTo(b, EqualTo(c, d)))` returns `true`
In conclusion, that will match the first one in the above two cases (not the second one), i.e. https://github.com/apache/spark/pull/33930/files/f5d3984c04029bdb7f4f00ef4de25a9e9fbc75cb#diff-d43484d56a4d9991066b5c00d12ec2465c75131e055fc02ee7fb6dfd45b5006fR472
After all it will optimize `Not(EqualTo(a, EqualTo(b, EqualTo(c, d))))` => `EqualTo(a, Not(EqualTo(b, EqualTo(c, d))))` => `EqualTo(a, EqualTo(b, Not(EqualTo(c, d))))` => `EqualTo(a, EqualTo(b, EqualTo(c, Not(d))))` => `EqualTo(a, EqualTo(b, EqualTo(c, e)))` where `e` is optimized `Not(d)`

Reply:
added test case for this https://github.com/apache/spark/blob/ef255953ccb043f473c35b5fb1d756e1d6a976b9/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala#L441